### PR TITLE
Feature/api to consumer lag

### PR DESCRIPTION
**Pull request by code-mad**

### CODE REVIEW

File: src/main/java/com/opendxl/streaming/client/Channel.java [hunk 1]
```diff
@@ -1246,6 +1247,124 @@ private String logConsumerId() {
...
+
+        StringBuilder api = new StringBuilder(consumerPathPrefix)
+                .append("/consumers/")
+                .append(consumerId)
+                .append("/updateFilter");
+
+        final String logMessage = new StringBuilder(logConsumerId())
+                .append(" to ").toString();
+        try {
+
```
Comment (umanekar-mcafee) [inline src/main/java/com/opendxl/streaming/client/Channel.java:1285]: Remove extra line.

File: src/main/java/com/opendxl/streaming/client/Channel.java [hunk 2]
```diff
@@ -1246,6 +1247,124 @@ private String logConsumerId() {
...
+            request.post(api.toString(), body, UPDATE_FILTER_ERROR_MAP);
+            if (logger.isDebugEnabled()) {
+                logger.debug("Update filter call " + logMessage);
+            }
+
```
Comment (umanekar-mcafee) [inline src/main/java/com/opendxl/streaming/client/Channel.java:1290]: Remove extra line.

File: src/main/java/com/opendxl/streaming/client/Channel.java [hunk 3]
```diff
@@ -1246,6 +1247,124 @@ private String logConsumerId() {
...
+            logger.error("Failed to update filter " + logMessage, error);
+            throw error;
+        }
+
+    }
+
+    public String callConsumerGroupDescribe(final String consumerGroupName)
+            throws PermanentError, TemporaryError {
+
+        acquireAndEnsureChannelIsActive();
```
Comment (umanekar-mcafee) [inline src/main/java/com/opendxl/streaming/client/Channel.java:1302]: Do we need this -   acquireAndEnsureChannelIsActive() ?

File: src/main/java/com/opendxl/streaming/client/Channel.java [hunk 4]
```diff
@@ -1246,6 +1247,124 @@ private String logConsumerId() {
...
+        if (consumerId == null || consumerId.isEmpty()) {
+            String error = "Consumer instance id can not be null or empty";
+            logger.error(error);
+            throw new PermanentError(error);
+        }
+
+        Gson gson = new Gson();
+        byte[] body = null;
+        if (null == filter || filter.isEmpty()) {
+            Topics topicsToBeSubscribed = new Topics(null);
```
Comment (umanekar-mcafee) [inline src/main/java/com/opendxl/streaming/client/Channel.java:1270]: Remove this Topics in request - as we discussed.

File: src/main/java/com/opendxl/streaming/client/Channel.java [hunk 5]
```diff
@@ -1246,6 +1247,124 @@ private String logConsumerId() {
...
+            if (logger.isDebugEnabled()) {
+                logger.debug("Update filter call " + logMessage);
+            }
+
+        } catch (final ClientError error) {
+            error.setApi("update-filter");
+            logger.error("Failed to update filter " + logMessage, error);
+            throw error;
+        }
+
```
Comment (umanekar-mcafee) [inline src/main/java/com/opendxl/streaming/client/Channel.java:1296]: Remove extra line

File: src/main/java/com/opendxl/streaming/client/Channel.java [hunk 6]
```diff
@@ -1246,6 +1247,124 @@ private String logConsumerId() {
...
+    }
+
+    public String callConsumerGroupDescribe(final String consumerGroupName)
+            throws PermanentError, TemporaryError {
+
```
Comment (umanekar-mcafee) [inline src/main/java/com/opendxl/streaming/client/Channel.java:1301]: Remove extra line.

File: src/main/java/com/opendxl/streaming/client/Channel.java [hunk 7]
```diff
@@ -1246,6 +1247,124 @@ private String logConsumerId() {
...
+        }
+
+        response = consumeLoopToDescribeConsumerGroup(consumerGroupName);
+        logger.info("Final response {}", response);
+        return response;
+    }
+
+    private String consumeLoopToDescribeConsumerGroup(final String consumerGroup)
+            throws PermanentError, TemporaryError {
+
```
Comment (umanekar-mcafee) [inline src/main/java/com/opendxl/streaming/client/Channel.java:1318]: Remove extra line.

File: src/main/java/com/opendxl/streaming/client/Channel.java [hunk 8]
```diff
@@ -1246,6 +1247,124 @@ private String logConsumerId() {
...
+        String response = null;
+        try {
+            response = callConsumerLagApi(consumerGroup);
+        } catch (ConsumerError e) {
+            logger.error("Exception ex {}", e);
+        }
+        return response;
+
```
Comment (umanekar-mcafee) [inline src/main/java/com/opendxl/streaming/client/Channel.java:1326]: Remove extra line.

File: src/main/java/com/opendxl/streaming/client/Channel.java [hunk 9]
```diff
@@ -1246,6 +1247,124 @@ private String logConsumerId() {
...
+     * @throws TemporaryError if the subscription attempt fails.
+     */
+    public String callConsumerLagApi(final String consumerGroup) throws ConsumerError, PermanentError, TemporaryError {
+        acquireAndEnsureChannelIsActive();
+        String consumerGroupResponse = null;
+        StringBuilder api = new StringBuilder()
+                .append("/v1/consumers/consumergroup/topic-wise-lag/")
+                .append(consumerGroup);
+        try {
+            logger.error("Request url :", api.toString());
```
Comment (umanekar-mcafee) [inline src/main/java/com/opendxl/streaming/client/Channel.java:1345]: why error logger here ?
I assume it should be debug.

File: src/main/java/com/opendxl/streaming/client/Channel.java [hunk 10]
```diff
@@ -1246,6 +1247,124 @@ private String logConsumerId() {
...
+            consumerGroupResponse = request.get(api.toString(), CREATE_ERROR_MAP);
+            logger.error("consumerGroupResponse {}", consumerGroupResponse);
+            if (logger.isDebugEnabled()) {
+                logger.debug("consumer group call with callConsumerLagApi" + api.toString());
+            }
+
```
Comment (umanekar-mcafee) [inline src/main/java/com/opendxl/streaming/client/Channel.java:1351]: Remove extra line.

File: src/main/java/com/opendxl/streaming/client/Channel.java [hunk 11]
```diff
@@ -1246,6 +1247,124 @@ private String logConsumerId() {
...
+        } catch (final ConsumerError error) {
+            error.setApi("consumerGroupLag");
+            logger.error("Failed to consumer-group " + consumerGroup, error);
```
Comment (umanekar-mcafee) [inline src/main/java/com/opendxl/streaming/client/Channel.java:1354]: error message - Lets change wordings as bellow.
Failed to fetch the lag for consumer group 'Group-Name'

**Review by umanekar-mcafee — CHANGES_REQUESTED** [review]

File: src/main/java/com/opendxl/streaming/client/Channel.java (hunk 3 above)
Comment (code-mad) [reply src/main/java/com/opendxl/streaming/client/Channel.java:1298]: In this case it is required as we are creating a new Channel for consumer-lag api if we don't have active channel this method will handle that.

**Review by umanekar-mcafee — APPROVED** [review]

**Pull request merged**